### DISCUSSION TITLE
Filter data automatically in reader_las()

### DISCUSSION THREAD

**Discussion by wiesehahn** [r-lidar/lasR#46] [General]
Following the discussion in https://github.com/r-lidar/lasR/issues/45 which has nothing to with the issue itself. 

The question was 
> Would it, just in theory, be possible to evaluate the pipeline on execution and then automatically apply an appropriate filter to read only necessary data?

Your last answer did not answer my question I think. 

As you state this pipeline 
```r
tri = triangulate()
dtm = rasterize(1, tri)
pipeline = reader_las(filter = keep_class(2)) + tri + dtm
```
is more efficient than this pipeline
```r
tri = triangulate(filter = keep_class(2))
dtm = rasterize(1, tri)
pipeline = reader_las() + tri + dtm
```
while they both achieve the same result. This is clear.

This might not be the best example but what I meant was that in theory I think it might be possible to write the second pipeline but achieve the same performance if we are aware which data is needed. This could be evaluated by `exec(pipeline, on = f)` which then applies the reader with minimal data transfer internally if no dedicated reader settings are given (in this case `reader_las(filter = keep_class(2))`).
I guess to implement this would be tedious (if not impossible), especially for more advanced pipelines.

Is this more clear?

**Comment by Jean-Romain** [comment]
Ok you would like the engine to guess that

```r
tri = triangulate(filter = keep_class(2))
dtm = rasterize(1, tri)
pipeline = reader_las() + tri + dtm
```

could be automatically optimized into 

```r
tri = triangulate()
dtm = rasterize(1, tri)
pipeline = reader_las(filter = keep_class(2)) + tri + dtm
```

I think it is absolutely impossible in the general case. `lidR` does it but `lidR` processes the equivalent of only one stage. Like in your example. In the general case with `n` stages it is excessively hard.

- The concept of stages in a pipeline that can be ordered in any order defined by the user is already complex to handle without any bug. As you can see you found many edge cases. It is ok, the software is maturing but I put a lot of effort in ensuring it works in all circumstances, in all orders. Of course I cannot test all the possible permutations of stages.
- To the concept of pipelines, `lasR` adds the concept of reusable and interacting stages. This adds another layer of complexity, These, for examples, are possible:

```r
tri = triangulate(filter = keep_class(2))
dtm = rasterize(1, tri)
norm = transform_with(dtm)
chm = rasterize(0.5, "max", filter = keep_first())
chm2 = pit_fill(dtm)
write = write_las(filter = "-keep_random_fraction 0.5")
pipeline = reader_las() + write + tri + dtm + norm + chm + chm2

# ----------------------

dtm = load_raster("data_provider_dtm.tif")
norm = transform_with(dtm)
chm = rasterize(0.5, "max")
chm2 = pit_fill(dtm)
pipeline = reader_las() + dtm + norm + chm + chm2
```

So let's take the first pipeline above. Can we guess what would be the best optimization? There is no optimization possible. We, as humans, are very good at solving these kind of problem. With more stages and a total of 100 filters I don't see how to solve it programmatically. It is just the best way to generate invalid outputs transparently.

**Reply by wiesehahn** [reply]
Ok, I kind of expected that it it not trivial at all. As the reading stage is quite computational expensive and the optimization step of filtering data in the reading stage can be easily missed I thought it might make sense if its possible to always achieve optimal processing speed here. 
While I dont say this is necessary or possible for all cases I am still not completely convinced its impossible as I think in your examples above all data is needed in both if I am correct. In 1 the write stage writes randomly selected points of the entire pointcloud (so if cannot choose random data while reading we need all). In 2 the height model is created from all data. But in the original example above it seems more easy to evaluate that dtm is dependent on tri, which in turn has a filter included to only use class 2. 

These theoretical features aside I have two related questions:
- The [documentation on filters](https://r-lidar.github.io/lasR/reference/filters.html) mentions boolean combination of filters "-filter_and", I am wondering what is the correct syntax to apply this and is it possible to use something like "-filter_or" which is not mentioned? Might make sense to include an example. 
- In lidR we can additionally use `select` to read only attributes of interest, this is not possible with lasR. Is it actually possible with LAS/LAZ files to read only certain attributes or is this data read but simply not kept in memory?

**Reply by Jean-Romain** [reply]
> I thought it might make sense if its possible to always achieve optimal processing speed here.

It is not totally impossible actually. Each filter could return a flag corresponding to the points actually needed. And we could combine all the flags from all the filters to create a *master* flag. In your first example, it would look like this:

1. stage 1 `reader_las()` needs `NOTHING` (it does what we tell it to do).
2. stage 2 `triangulate()` needs `GROUND`.
3. stage 3 `rasterize()` needs `NOTHING` (it takes what stage 2 needs).

All combined we need `GROUND`.

My example is more complex but in practice it is simpler:

1. stage 1 `reader_las()` needs `NOTHING` (it does what we tell it to do).
2. stage 2 `write_las()` needs `ALL`.

All combined we need `ALL`. No need to analyze the entire pipeline.

I'm using `LASlib`, the library developed by LAStools, to read and write LAS and LAZ files. The filters come from `LASlib`. This would require significant modification in `LASlib` itself. This is definitely not something I can do within a week in my spare time. But it is something I could consider in the long term if someone shows me significant use cases where it would really bring improvements.

> The [documentation on filters](https://r-lidar.github.io/lasR/reference/filters.html) mentions boolean combination of filters "-filter_and", I am wondering what is the correct syntax to apply this and is it possible to use something like "-filter_or" which is not mentioned? Might make sense to include an example.

I'm using `LASlib` the library developed by LAStools to read and write LAS ans LAZ files. Sometime I do no know all the features myself. I'm not sure how these features are working and are poorly documented in LAStools itself. I don't know if `filter_and` is even useful because `-keep_first -keep_class 2` is already an `AND` operation.

> In lidR we can additionally use select to read only attributes of interest, this is not possible with lasR. Is it actually possible with LAS/LAZ files to read only certain attributes or is this data read but simply not kept in memory?

No, it is impossible in lasR for good reasons.  I'm not going to list them all, but an important one is that lidR does read all the attributes (it is not possible to do otherwise) but it does not expose everything into R to save memory. R is bad at memory management because it has on 2 types (`int` and `double`). So, I added `select` to mitigate the problem. The consequences is that you may write a point cloud with all the attributes zeroed if you do something like:

```r
las = readLAS("file.las", select = "xyz")
writeLAS(las, "output.las")
```

The `select` argument is a trade of and is disabled in functions like `normalized_height(ctg, ...)` to ensure users get a point cloud in output that is equivalent to the input. `lasR` is a pure C++ program decoupled from R. There is no problem of memory management. In lasR a point from a file format 1 takes 28 bytes of memory has described in the specifications. In lidR, when loaded in a `data.frame` using R's vectors the same points takes 68 bytes of memory (x2.5). Maybe even more.

 In `lasR` we always have all the attributes and I'm not going to change that. That would require deep changes in `LASlib` to add the same issues we have in `lidR`.

**Reply by wiesehahn** [reply]
> This would require significant modification in LASlib itself.

Could this not be evaluated on the lasR side, to me it sounds like its "just" about passing the correct filter to the reader?

> I'm using LASlib the library developed by LAStools to read and write LAS ans LAZ files. Sometime I do no know all the features myself. I'm not sure how these features are working and are poorly documented in LAStools itself. I don't know if filter_and is even useful because -keep_first -keep_class 2 is already an AND operation.

I just tested it and we can do `-keep_first -keep_class 2 -filter_and`, and also `-keep_first -keep_class 2 -filter_or`. Actually the equivalent of `-keep_first -keep_class 2` is `-keep_first -keep_class 2 -filter_or` by which we get all points which are EITHER first returns OR ground points. By `-keep_first -keep_class 2 -filter_and` we get all points which are first returns AND classified as ground.  

> In lasR we always have all the attributes and I'm not going to change that. That would require deep changes in LASlib to add the same issues we have in lidR.

Its a totally different topic and I am not an expert with that, but I am really curious if columnar formats such as Geoparquet / Geoarrow could be useful for pointcloud data in the future. This could allow to read only parts of data (spatial queries + attribute subset). Along with some other potential benefits such as zero-copy shared memory (arrow) to interoperate with data in multiple applications. In case you ever have any knowledge/opinions on this or good resources on how the LAS format is structured internally I would be interested.

**Reply by Jean-Romain** [reply]
>  it sounds like its "just" about passing the correct filter to the reader?

It is. But evaluating what is the correct filter is challenging in the general case.

> Its a totally different topic and I am not an expert with that, but I am really curious if columnar formats such as Geoparquet / Geoarrow could be useful for pointcloud data in the future.

I have no knowledge about these formats. Lot of things are possible but remember that I created `rlas`, `lidR`, and `lasR` alone in addition to my work. At some point it is no longer only a matter of existing technologies. Also, I have never seen any of these formats. Everybody I heard of is using `las/laz` files, at least in the academic world.

> This could allow to read only parts of data (spatial queries + attribute subset).

 `.las/.laz` files are already spatially indexed with `.lax` files and unlike `lidR`, `lasR` forces you to have spatially indexed files. `.copc.laz` are also supported.

> . In case you ever have any knowledge/opinions on this or good resources on how the LAS format is structured internally I would be interested.

The specifications are open https://www.asprs.org/wp-content/uploads/2019/07/LAS_1_4_r15.pdf
